src/Compiler/Utilities/TypeHashing.fs

```diff
@@ -357,68 +358,112 @@ module HashTastMemberOrVals =
-
-    let inline hashStamp (x: Stamp) : Hash = uint x * 2654435761u |> int
-
-    // The idea is to keep the illusion of immutability of TType.
-    // This hash must be stable during compilation, otherwise we won't be able to find keys or evict from the cache.
-    let rec hashTType ty : Hash =
+/// <summary>
+/// StructuralUtilities: produce a conservative structural fingerprint of <c>TType</c>.
+///
+/// Current (sole) usage:
+///   Key in the typeSubsumptionCache. The key must never give a false positive
+///   (two non-subsuming types producing identical token sequences). False negatives
+///   are acceptable and simply reduce cache hit rate.
+///
+/// Properties:
+///   * Uses per-compilation stamps (entities, typars, anon records, measures).
+///   * Emits shape for union cases (declaring type stamp + case name), tuple structness,
+///     function arrows, forall binders, nullness, measures, generic arguments.
+///   * Unknown/variable nullness => NeverEqual token to force inequality (avoid unsound hits).
+///
+/// Non-goals:
+///   * Cross-compilation stability.
+///   * Perfect canonicalisation or alpha-equivalence collapsing.
+///
+/// </summary>
+module StructuralUtilities =
+    [<Struct; CustomEquality; NoComparison>]
+    type NeverEqual =
+        struct
+            interface System.IEquatable<NeverEqual> with
+                member _.Equals _ = false
+
+            override _.Equals _ = false
+            override _.GetHashCode() = 0
+        end
+
+        static member Singleton = NeverEqual()
+
+    [<Struct; NoComparison; RequireQualifiedAccess>]
+    type TypeToken =
+        | Stamp of stamp: Stamp
+        | UCase of name: string
+        | Nullness of nullness: NullnessInfo
+        | TupInfo of b: bool
+        | MeasureOne
+        | MeasureRational of rational: Rational
+        | NeverEqual of never: NeverEqual
+
+    type TypeStructure = TypeToken[]
+
+    [<Literal>]
+    let private initialTokenCapacity = 4
+
+    let inline toNullnessToken (n: Nullness) =
+        match n with
+        | Nullness.Known k -> TypeToken.Nullness k
+        // If nullness is not known we must treat the types as not equal for caching purposes.
+        | Nullness.Variable _ -> TypeToken.NeverEqual NeverEqual.Singleton
+
+    let rec private accumulateMeasure (tokens: ResizeArray<TypeToken>) (m: Measure) =
+        match m with
+        | Measure.Var mv -> tokens.Add(TypeToken.Stamp mv.Stamp)
+        | Measure.Const(tcref, _) -> tokens.Add(TypeToken.Stamp tcref.Stamp)
+        | Measure.Prod(m1, m2, _) ->
+            accumulateMeasure tokens m1
+            accumulateMeasure tokens m2
+        | Measure.Inv m1 -> accumulateMeasure tokens m1
+        | Measure.One _ -> tokens.Add(TypeToken.MeasureOne)
+        | Measure.RationalPower(m1, r) ->
+            accumulateMeasure tokens m1
+            tokens.Add(TypeToken.MeasureRational r)
+
+    let rec private accumulateTType (tokens: ResizeArray<TypeToken>) (ty: TType) =
```

Comment:
So the `tokens` here is a depth-first flattened list of anything `potentially interesting for equality`?

I think that this is safe against infinite looping since the types are stripped, and since `TType_var` handler does not do a recursive call 👍